server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -106,11 +115,61 @@ private static void internalParseDocument(RootObjectMapper root, MetadataFieldMa
             parseObjectOrNested(context, root);
         }
 
+        executeIndexTimeScripts(context);
+
         for (MetadataFieldMapper metadataMapper : metadataFieldsMappers) {
             metadataMapper.postParse(context);
         }
     }
 
+    private static void executeIndexTimeScripts(ParseContext context) {
```

Comment:
I find this approach clearer than before, thanks! One thing I still notice is that a Script becomes a MapperScript, which becomes an IndexTimeScript which becomes a Consumer<LeafReaderContext>. Conceptually the same thing is represented in three different ways. I think I see the distinction between MapperScript and IndexTimeScript, as the former is executed against a real index to fetch values as well (could we do something to highlight that?). I am thinking that the OneTimeFieldExecutor could implement IndexTimeScript too, so that at least one step can be avoided. We do wrap the script but it is still something of the same class, which I find easier to reason about.

Reply:
I've folded OneTimeFieldExecutor directly into the consumer for DocumentLeafReader - is this any clearer?